File: web/src/lib/types/indexer.ts
export interface Indexer {
  id: number
  name: string
  type: "newznab" | "torznab"
  baseUrl: string
  apiPath: string
  apiKey: string
  categories: string
  priority: number
  enabled: boolean
  enableRss: boolean
  enableAutomaticSearch: boolean
  enableInteractiveSearch: boolean
  additionalParameters: string
  minimumSeeders: number
  seedRatio?: number | null
  seedTime?: number | null
  createdAt: string
  updatedAt: string
}

export interface CreateIndexerInput {
  name: string
  type: "newznab" | "torznab"
  baseUrl: string
  apiPath?: string
  apiKey: string
  categories?: string
  priority?: number
  enabled?: boolean
  enableRss?: boolean
  enableAutomaticSearch?: boolean
  enableInteractiveSearch?: boolean
  additionalParameters?: string
  minimumSeeders?: number
  seedRatio?: number | null
  seedTime?: number | null
}

export interface UpdateIndexerInput extends CreateIndexerInput {
  id?: number
}

export interface SearchResult {
  guid: string
  title: string
  description?: string
  size: number
  pubDate: string
  category?: string
  categoryId?: string
  downloadUrl: string
  infoUrl?: string
  comments?: number
  seeders?: number
  leechers?: number
  grabs?: number
  quality?: string
  qualityRank?: number
  indexerId: number
  indexerName: string
}

import type { ColumnConfig } from "./search"

export interface SearchResponse {
  results: SearchResult[]
  total: number
  columnConfig?: ColumnConfig
}

export interface TestIndexerResponse {
  success: boolean
  message: string
}

export interface IndexerOptions {
  minimumAge: number // Minutes
  retention: number // Days (0 = unlimited)
  maximumSize: number // MB (0 = unlimited)
  rssSyncInterval: number // Minutes (0 = disabled)
  preferIndexerFlags: boolean
  availabilityDelay: number // Days
  updatedAt: string
}

export type UpdateIndexerOptionsInput = Omit<IndexerOptions, "updatedAt">

File: internal/search/indexer.go
package search

import (
	"context"
	"time"
)

// Indexer defines the interface for search indexers (Newznab, Torznab).
type Indexer interface {
	Name() string
	Type() string
	Search(ctx context.Context, query SearchQuery) ([]Result, error)
	Caps(ctx context.Context) (*Capabilities, error)
	Test(ctx context.Context) error
}

// SearchQuery represents a search request.
type SearchQuery struct {
	Query    string
	Author   string
	Title    string
	ISBN     string
	Category []string
	Limit    int
	Offset   int
}

// Result represents a single search result from an indexer.
type Result struct {
	GUID        string    `json:"guid"`
	Title       string    `json:"title"`
	Description string    `json:"description,omitempty"`
	Size        int64     `json:"size"`
	PubDate     time.Time `json:"pubDate"`
	Category    string    `json:"category,omitempty"`
	CategoryID  string    `json:"categoryId,omitempty"`
	DownloadURL string    `json:"downloadUrl"`
	InfoURL     string    `json:"infoUrl,omitempty"`
	Comments    int       `json:"comments,omitempty"`
	Seeders     int       `json:"seeders,omitempty"`
	Leechers    int       `json:"leechers,omitempty"`
	Grabs       int       `json:"grabs,omitempty"`
	Quality     string    `json:"quality,omitempty"`
	QualityRank int       `json:"qualityRank,omitempty"`
	IndexerID   int64     `json:"indexerId"`
	IndexerName string    `json:"indexerName"`
}

// Capabilities describes what an indexer can do.
type Capabilities struct {
	Searching struct {
		Search     bool `json:"search"`
		BookSearch bool `json:"bookSearch"`
	} `json:"searching"`
	Categories []Category `json:"categories"`
}

// Category represents a category supported by an indexer.
type Category struct {
	ID          string     `json:"id"`
	Name        string     `json:"name"`
	SubCategory []Category `json:"subCategory,omitempty"`
}

// IndexerConfig holds the configuration for an indexer.
type IndexerConfig struct {
	ID                      int64    `json:"id"`
	Name                    string   `json:"name"`
	Type                    string   `json:"type"`
	BaseURL                 string   `json:"baseUrl"`
	APIPath                 string   `json:"apiPath"`
	APIKey                  string   `json:"apiKey"`
	Categories              string   `json:"categories"`
	Priority                int      `json:"priority"`
	Enabled                 bool     `json:"enabled"`
	EnableRSS               bool     `json:"enableRss"`
	EnableInteractiveSearch bool     `json:"enableInteractiveSearch"`
	AdditionalParameters    string   `json:"additionalParameters"`
	MinimumSeeders          int      `json:"minimumSeeders"`      // Torznab only
	SeedRatio               *float64 `json:"seedRatio,omitempty"` // Torznab only, nil = use client default
	SeedTime                *int     `json:"seedTime,omitempty"`  // Torznab only, minutes, nil = use client default
	CreatedAt               string   `json:"createdAt"`
	UpdatedAt               string   `json:"updatedAt"`
}

// IndexerOptions holds global indexer settings.
type IndexerOptions struct {
	MinimumAge         int    `json:"minimumAge"`         // Minutes (Usenet: min age before grab)
	Retention          int    `json:"retention"`          // Days (Usenet: 0 = unlimited)
	MaximumSize        int    `json:"maximumSize"`        // MB (0 = unlimited)
	RSSSyncInterval    int    `json:"rssSyncInterval"`    // Minutes (0 = disabled)
	PreferIndexerFlags bool   `json:"preferIndexerFlags"` // Prioritize releases with special flags
	AvailabilityDelay  int    `json:"availabilityDelay"`  // Days
	UpdatedAt          string `json:"updatedAt"`
}
